src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Formatting/LspCSharpOnTypeFormattingPass.cs

```diff
@@ -0,0 +1,39 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.Language;
+using Microsoft.AspNetCore.Razor.LanguageServer.CodeActions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor.DocumentMapping;
+using Microsoft.CodeAnalysis.Razor.Formatting;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.Formatting;
+
+internal sealed class LspCSharpOnTypeFormattingPass(
+    IDocumentMappingService documentMappingService,
+    ILoggerFactory loggerFactory)
+    : CSharpOnTypeFormattingPassBase(documentMappingService, loggerFactory)
+{
+
+    protected override async Task<TextEdit[]> AddUsingStatementEditsIfNecessaryAsync(CodeAnalysis.Razor.Formatting.FormattingContext context, RazorCodeDocument codeDocument, SourceText csharpText, TextEdit[] textEdits, SourceText originalTextWithChanges, TextEdit[] finalEdits, CancellationToken cancellationToken)
+    {
+        if (context.AutomaticallyAddUsings)
+        {
+            // Because we need to parse the C# code twice for this operation, lets do a quick check to see if its even necessary
+            if (textEdits.Any(e => e.NewText.IndexOf("using") != -1))
+            {
+                var usingStatementEdits = await AddUsingsCodeActionProviderHelper.GetUsingStatementEditsAsync(codeDocument, csharpText, originalTextWithChanges, cancellationToken).ConfigureAwait(false);
```

Comment:
From a quick look at `AddUsingsCodeActionProviderHelper` there doesn't seem to be anything in it that couldn't be moved down to a lower level, which would avoid the abstract base class for this pass etc. Am I missing something subtle?

Reply:
I made an attempt at moving that one class, but it does have dependencies on a few other things in CodeActions. I'd rather not go down the path of moving a large chunk of code actions as well. It might very well be that that class will end up not being abstract long-term, but seems OK short-term to just help stop migration to the common layer at that point.